pkg/kubelet/kubelet.go

```diff
@@ -2276,32 +2277,73 @@ func (kl *Kubelet) rejectPod(pod *api.Pod, reason, message string) {
+// The *api.Node is obtained as follows:
+// 1. When kubelet is in standaloneMode, manufactures *api.Node with kubelet
+// 2. Try to get *api.Node from kubelet's nodeInfo, if failed, we try 3.
+// 3. Try to get *api.Node from apiserver
+func (kl *Kubelet) getNodeAnyWay() (*api.Node, error) {
+	var node *api.Node
+	if kl.standaloneMode {
+		if n, err := kl.initialNodeStatus(); err == nil {
+			return n, nil
+		}
+	}
+	if n, err := kl.nodeInfo.GetNodeInfo(kl.nodeName); err == nil {
+		return n, nil
+	}
+	node, err := kl.kubeClient.Core().Nodes().Get(kl.nodeName)
+	if err != nil {
+		return node, err
+	}
+	return node, nil
+}
+
 // canAdmitPod determines if a pod can be admitted, and gives a reason if it
 // cannot. "pod" is new pod, while "pods" include all admitted pods plus the
 // new pod. The function returns a boolean value indicating whether the pod
 // can be admitted, a brief single-word reason and a message explaining why
 // the pod cannot be admitted.
-//
-// This needs to be kept in sync with the scheduler's and daemonset's fit predicates,
-// otherwise there will inevitably be pod delete create loops. This will be fixed
-// once we can extract these predicates into a common library. (#12744)
 func (kl *Kubelet) canAdmitPod(pods []*api.Pod, pod *api.Pod) (bool, string, string) {
-	if hasHostPortConflicts(pods) {
-		return false, "HostPortConflict", "cannot start the pod due to host port conflict."
-	}
-	if !kl.matchesNodeSelector(pod) {
-		return false, "NodeSelectorMismatching", "cannot be started due to node selector mismatch"
-	}
-	cpu, memory := kl.hasInsufficientfFreeResources(pods)
-	if cpu {
-		return false, "InsufficientFreeCPU", "cannot start the pod due to insufficient free CPU."
-	} else if memory {
-		return false, "InsufficientFreeMemory", "cannot be started due to insufficient free memory"
+	node, err := kl.getNodeAnyWay()
+	if err != nil {
+		fmt.Errorf("Cannot get Node info, %v", err)
 	}
+	otherPods := []*api.Pod{}
+	for _, p := range pods {
+		if p != pod {
+			otherPods = append(otherPods, p)
+		}
+	}
+	nodeInfo := schedulercache.CreateNodeNameToInfoMap(otherPods)[kl.nodeName]
+	fit, err := predicates.RunGeneralPredicates(pod, kl.nodeName, nodeInfo, node)
+	var reason string
+	var message string
+	if !fit {
+		if re, ok := err.(*predicates.PredicateFailureError); ok {
+			reason = re.PredicateName
+			message = re.Error()
+			glog.Infof("predicate fails on Pod: %v, for reason: %v", pod.Name, message)
+			return fit, reason, message
+		}
+		if re, ok := err.(*predicates.InsufficientResourceError); ok {
+			reason = re.ResourceName
+			message = re.Error()
+			glog.Infof("predicate fails on Pod: %v, for reason: %v", pod.Name, message)
+			return fit, reason, message
+		}
+		reason = "Unexpected predicate failure type"
+		message = fmt.Sprintf("GeneralPredicates failed due to %v, which is unexpected.\n", err)
+		glog.Warningf("failed to admit pod %q - %s", pod.Name, message)
```

Comment:
s/failed/Failed

nit: instead of using `pod.Name`, just call `format.Pod(pod))`, which includes pod name/namespace and uid.